### PR TITLE
Diffusive Data Assimilation v1 with USGS streamflow data

### PR DESCRIPTION
**Pull request by kumdonoaa**
[Short description explaining the high-level reason for the pull request]

## Additions

- streamflow DA capability to diffusive.f90 (When USGS streamflow data is available to a particular stream reach, the data instead of estimated value is used as streamflow boundary condition at the bottom end of the reach for computing streamflow of all upstream stream segments within the reach at the next simulation time step)
- Three new input arguments, i.e., nts_da_g, usgs_da_g, usgs_da_reach_g, were created inside diffusive_utils.py and passed to diffusive.f90.  nts_da_g = the number of DA data time steps;  usgs_da_g =  USGS streamflow data at available stream reach;  usgs_da_reach_g = the indices of stream reaches that  are co-located with UGSG streamflow gages.
- New function def fp_da_map in diffusive_utils.py was created to produce the three arguments.
- In Test_AnA.yaml, "diffusive_streamflow_DA" variable was added as a switch to turn on or off diffusive streamflow DA.
- Test_AnA_ncar.yaml file is added to simulate a period from 2021-09-01 up to 28 hrs.

## Removals

-

## Changes

-

## Testing

1.  Diffusive DA (Data Assimilation) at Lower Colorado River with USGS streamflow data at 7 locations over 28 hrs
As specified in test_AnA.yaml, seven USGS observed streamflow time series over 28 hrs were used for discharge Data Assimilation.  Each of the time series replaced estimated streamflow values for next simulation time steps at the bottom computation node of the corresponding stream reach.  When given USGS data is in poor quality or missing, the estimated values are plugged back in.  Then, the impact of streamflow values at the bottom node is propagated upstream in computing streamflow at upstream nodes within the same reach.   The figure below compares DA and non-DA streamflows over 28 hrs with 5 min time steps.

diffusive DA prototype1

2.  No-DA Diffusive run with NCAR tested time period (from 2021-09-01 up to 28 hrs)
Based on inland_config_hybrid_old.yaml, all necessary forcing and computation parameters were copied and pasted to test_AnA_ncar.yaml.  Also, diffusive.f90 was updated so that there is no initial hiccup in the diffusive simulation. (See the update in https://github.com/kumdonoaa/t-route/blob/29cbdc54e9ebeea5d31dbb7e4332fd9c04169a7b/src/kernel/diffusive/diffusive.f90#L554-L559).  The figure below shows the results from this run.

Recording 2022-04-18 at 16 37 17 (1)

  
## Notes

-

## Todos

-

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

**Comment by awlostowski-noaa** [comment]
great to see this on the docket. Let's review together, soon.

File: test/LowerColorado_TX/test_AnA_ncar.yaml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+# $ python -m nwm_routing -f -V3 test_AnA.yaml
```
Comment (awlostowski-noaa) [inline test/LowerColorado_TX/test_AnA_ncar.yaml:1]: Please do not commit this file.

**Review by awlostowski-noaa — COMMENTED** [review]
There are a few excessive files committed here that we would not want to merge.
- test_AnA_ncar.yaml: This file is only relevant to running a test on Cheyenne and some users wouldn't have access. Plus committing files with full directory paths is a security risk
- sine_tide_matagorda_city_...: this appears to be for development testing purposes and it is data. No need to commit it into the repository, unless you want it to be part of the canned Lower Colorado test case that ships with the repository. 
- diffusive_natural_xs_... : See recent pull request, a cross section file has already been merged. Please rebase this branch.

**Comment by awlostowski-noaa** [comment]
Your first animation seems to show that diffusive DA causes some pretty extreme flow discontinuities that are suspect. For example @ x = 170000. What is the cause of this? Is there a gauge at this location, or a gauge just downstream of this location? On one hand, I would expect gauge data to impose discontinuities in the longitudinal flow profile. But on the other hand, I wouldn't expect to see such drastic discontinuities along a profile containing seven gauges. Is it possible that you are substituting data from a gauge into the wrong spatial location?

File: test/LowerColorado_TX/test_AnA.yaml [hunk 2]
```diff
@@ -57,6 +59,7 @@ compute_parameters:
         streamflow_da:
             #----------
             streamflow_nudging       : True
+            diffusive_streamflow_DA  : False
```
Comment (awlostowski-noaa) [inline test/LowerColorado_TX/test_AnA.yaml:62]: maybe change the name of this parameter to `diffusive_streamflow_nudging` to be more in-line with naming convention.

File: test/LowerColorado_TX/test_AnA.yaml [hunk 3]
```diff
@@ -38,8 +38,10 @@ compute_parameters:
         wrf_hydro_waterbody_ID_crosswalk_file    : domain/LAKEPARM.nc
         wrf_hydro_waterbody_crosswalk_filter_file: domain/RouteLink.nc
     hybrid_parameters:
-        run_hybrid_routing: False
+        run_hybrid_routing: True
```
Comment (awlostowski-noaa) [inline test/LowerColorado_TX/test_AnA.yaml:41]: I'd prefer if we didn't change the parameter values in the testing yaml file. The original yaml file for the Lower Colorado test case is initially set up to run a purely MC simulation with streamflow nudging and reservoir DA.

File: test/LowerColorado_TX/test_AnA.yaml [hunk 4]
```diff
@@ -38,8 +38,10 @@ compute_parameters:
         wrf_hydro_waterbody_ID_crosswalk_file    : domain/LAKEPARM.nc
         wrf_hydro_waterbody_crosswalk_filter_file: domain/RouteLink.nc
     hybrid_parameters:
-        run_hybrid_routing: False
+        run_hybrid_routing: True
         diffusive_domain  : domain/diffusive_domain.yaml
+        use_natl_xsections: True
+        topobathy_domain  : input/diffusive_natural_xs_20220331.nc
```
Comment (awlostowski-noaa) [inline test/LowerColorado_TX/test_AnA.yaml:44]: See recently merged PR - these parameters have already been added, as well as a natural cross section file. Please rebase.

File: test/LowerColorado_TX/test_AnA.yaml [hunk 5]
```diff
@@ -29,7 +29,7 @@ compute_parameters:
     compute_kernel         : V02-structured
     assume_short_ts        : True
     subnetwork_target_size : 10000
-    cpu_pool               : 36
+    cpu_pool               : 1
```
Comment (awlostowski-noaa) [inline test/LowerColorado_TX/test_AnA.yaml:32]: Why are we changing this?

File: src/troute-routing/troute/routing/diffusive_utils.py [hunk 6]
```diff
@@ -502,6 +502,88 @@ def fp_naturalxsec_map(
...
+    tfin_g          -- (float) diffusive model's final simulation time [hr] 
+    
+    Returns
+    -------
+    nts_da_g        -- (int) number of DA timesteps
+    usgs_da_g       -- (float) usgs oberved streamflow data [cms]
+    usgs_da_reach_g -- (int) indices of stream reaches where DA is applied    
+    """
+    from datetime import datetime, timedelta
+    import pandas as pd
```
Comment (awlostowski-noaa) [inline src/troute-routing/troute/routing/diffusive_utils.py:536]: not good practice to import libraries in functions. This should be at the top of the script. Hard to believe that we don't already have pandas imported, here?

File: src/troute-routing/troute/routing/diffusive_utils.py [hunk 7]
```diff
@@ -502,6 +502,88 @@ def fp_naturalxsec_map(
...
+                #    segID= seg_list[seg]
+                for tsi in range (0,nts_da_g):
+                    t_min= dt_da_g/60.0*float(tsi)                        
+                    dmy1= usgs_da_g[tsi, frj]
+                    usgs_da.write("%s %s %s\n" %\
+                                (str(frj+1), str(t_min), str(dmy1)) )                   
+                                # <- +1 is added to frj for accommodating Fortran indexing. 
+    # test
+    np.savetxt("./output/usgs_da_reach_g.txt", usgs_da_reach_g, fmt='%15.5f')
+    '''
```
Comment (awlostowski-noaa) [inline src/troute-routing/troute/routing/diffusive_utils.py:584]: Is there good reason for keeping this code and merging it into master?

File: src/troute-routing/troute/routing/diffusive_utils.py [hunk 8]
```diff
@@ -541,15 +625,17 @@ def diffusive_input_data_v02(
...
     """
-    
+
     # lateral inflow timestep (sec)
     dt_ql_g = dt * qts_subdivisions
     # upstream boundary condition timestep (sec)
     dt_ub_g = dt
     # downstream boundary condition timestep (sec)
-    dt_db_g = dt * qts_subdivisions
+    dt_db_g = 360.0 # dt * qts_subdivisions
```
Comment (awlostowski-noaa) [inline src/troute-routing/troute/routing/diffusive_utils.py:634]: hard-wiring values is risky. What happens if this changes? Also, the downstream boundary data has a 6-minute time interval? That seems quite arbitrary.

File: src/troute-routing/troute/routing/diffusive_utils.py [hunk 9]
```diff
@@ -541,15 +625,17 @@ def diffusive_input_data_v02(
...
     # tributary inflow timestep (sec)
     dt_qtrib_g = dt
+    # usgs_df time step used for data assimilation. The original timestep of USGS streamflow is 15 min but later interpolated at every dt in min.
+    dt_da_g = dt # [sec]
```
Comment (awlostowski-noaa) [inline src/troute-routing/troute/routing/diffusive_utils.py:638]: Do we need a new parameter for this, or can we just operate under the assumption that the USGS data will be at time resolution dt?

File: src/troute-routing/troute/routing/diffusive_utils.py [hunk 10]
```diff
@@ -715,7 +802,16 @@ def diffusive_input_data_v02(
...
+    
+    np.savetxt("./output/frnw_ar.txt", frnw_g, fmt='%10i')
+    frj= -1
+    with open("./output/pynw.txt",'a') as pynwtxt:    
+        for x in range(mx_jorder,-1,-1):   
+            for head_segment, reach in ordered_reaches[x]:       
+                frj= frj+1     
+                pynwtxt.write("%s %s\n" %\
+                           (str(frj+1), str(pynw[frj]))) 
+
```
Comment (awlostowski-noaa) [inline src/troute-routing/troute/routing/diffusive_utils.py:814]: what is this?

**Review by awlostowski-noaa — COMMENTED** [review]
additional remarks

File: test/LowerColorado_TX/test_AnA.yaml (hunk 2 above)
Comment (kumdonoaa) [reply test/LowerColorado_TX/test_AnA.yaml:62]: Got it.

File: test/LowerColorado_TX/test_AnA.yaml (hunk 4 above)
Comment (kumdonoaa) [reply test/LowerColorado_TX/test_AnA.yaml:44]: ok.

File: test/LowerColorado_TX/test_AnA.yaml (hunk 5 above)
Comment (kumdonoaa) [reply test/LowerColorado_TX/test_AnA.yaml:32]: just for test. Will be back to the original value.

**Pull request closed without merging**